### PR TITLE
p4c-of: Support "optional" match type.

### PR DESCRIPTION
**Pull request by blp**

### CODE REVIEW

File: ofp4/backend.cpp [hunk 1]
```diff
@@ -625,6 +622,86 @@ class FlowGenerator : public Inspector {
...
+            if (matchType->path->name.name == "optional") {
+                // For an optional field, we need a flow for None and a flow
+                // for Some.  The flow for None doesn't have a match component;
+                // add it first.
+                std::vector<cstring> args;
+                tableArgs.push_back(new IR::DDlogConstructorExpression(IR::ID("None"), args));
+                convertKey(table, tableCases, tableArgs, match, curKey + 1, end, nKeys);
+                tableArgs.pop_back();
+
+                // Then discard the None and add the Some.  The match component
```
Comment (mbudiu-vmw) [inline ofp4/backend.cpp:650]: should this comment be a bit earlier?

File: ofp4/backend.cpp [hunk 2]
```diff
@@ -625,6 +622,86 @@ class FlowGenerator : public Inspector {
         declarations->push_back(makeFlowRule(flowRule, cfgtable->table->externalName()));
     }
 
+    void convertKey(CFG::TableNode* table,
+                    IR::Vector<IR::DDlogMatchCase>* tableCases,
```
Comment (mbudiu-vmw) [inline ofp4/backend.cpp:626]: this seems to be an "out" argument, maybe you want to add a comment.
I think this function deserves a comment too, ideally describing the parameters.

**Review by mbudiu-vmw — APPROVED** [review]
I have to confess that I have paged out the structure of this code.
It seems to be fine, though. Some comments would help.

File: ofp4/backend.cpp (hunk 1 above)
Comment (blp) [reply ofp4/backend.cpp]: Thanks. I revised the comments.

File: ofp4/backend.cpp (hunk 2 above)
Comment (blp) [reply ofp4/backend.cpp]: You are right. This needs some comments. I added some. They probably help.

**Pull request merged**